File: src/main/kotlin/jp/trap/plutus/pteron/features/stats/service/StatsUpdateJob.kt
package jp.trap.plutus.pteron.features.stats.service

import jp.trap.plutus.pteron.common.domain.UnitOfWork
import jp.trap.plutus.pteron.common.domain.model.AccountId
import jp.trap.plutus.pteron.common.domain.model.ProjectId
import jp.trap.plutus.pteron.common.domain.model.UserId
import jp.trap.plutus.pteron.features.account.domain.gateway.EconomicGateway
import jp.trap.plutus.pteron.features.account.domain.model.Account
import jp.trap.plutus.pteron.features.project.domain.model.Project
import jp.trap.plutus.pteron.features.project.domain.repository.ProjectRepository
import jp.trap.plutus.pteron.features.stats.domain.model.*
import jp.trap.plutus.pteron.features.stats.domain.repository.StatsCacheRepository
import jp.trap.plutus.pteron.features.transaction.domain.repository.TransactionRepository
import jp.trap.plutus.pteron.features.user.domain.model.User
import jp.trap.plutus.pteron.features.user.domain.repository.UserRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import org.koin.core.annotation.Single
import org.slf4j.LoggerFactory
import kotlin.time.Clock
import kotlin.time.Duration.Companion.hours
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.minutes

@Single
class StatsUpdateJob(
    private val cacheRepository: StatsCacheRepository,
    private val transactionRepository: TransactionRepository,
    private val economicGateway: EconomicGateway,
    private val userRepository: UserRepository,
    private val projectRepository: ProjectRepository,
    private val unitOfWork: UnitOfWork,
) {
    private val logger = LoggerFactory.getLogger(StatsUpdateJob::class.java)

    fun start(scope: CoroutineScope) {
        scope.launch {
            delay(10_000.milliseconds)

            while (isActive) {
                try {
                    updateAllStats()
                } catch (e: Exception) {
                    logger.error("Failed to update stats cache", e)
                }
                delay(1.minutes)
            }
        }
    }

    suspend fun updateAllStats() {
        logger.info("Starting stats cache update...")
        val startTime = System.currentTimeMillis()

        val users = unitOfWork.runInTransaction { userRepository.findAll() }
        val projects = unitOfWork.runInTransaction { projectRepository.findAll() }
        val accountIds = (users.map { it.accountId } + projects.map { it.accountId }).distinct()
        val accountBalanceMap = economicGateway.findAccountsByIds(accountIds).associateBy { it.accountId }

        for (term in Term.entries) {
            try {
                unitOfWork.runInTransaction {
                    updateSystemStats(term, users, projects, accountBalanceMap)
                    updateUsersAggregateStats(term, users, accountBalanceMap)
                    updateProjectsAggregateStats(term, projects, accountBalanceMap)

                    val userRankingsCurrentTerm = calculateAllUserRankings(term, users, accountBalanceMap)
                    val userRankingsPreviousTerm = calculateAllUserRankings(term, users, accountBalanceMap, previous = true)
                    val projectRankingsCurrentTerm = calculateAllProjectRankings(term, projects, accountBalanceMap)
                    val projectRankingsPreviousTerm =
                        calculateAllProjectRankings(term, projects, accountBalanceMap, previous = true)

                    for (rankingType in RankingType.entries) {
                        updateUserRankings(term, rankingType, userRankingsCurrentTerm, userRankingsPreviousTerm)
                        updateProjectRankings(
                            term,
                            rankingType,
                            projectRankingsCurrentTerm,
                            projectRankingsPreviousTerm,
                        )
                    }
                }
            } catch (e: Exception) {
                logger.error("Failed to update stats for term ${term.key}", e)
            }
        }

        val elapsed = System.currentTimeMillis() - startTime
        logger.info("Stats cache update completed in ${elapsed}ms")
    }

    private suspend fun updateSystemStats(
        term: Term,
        users: List<User>,
        projects: List<Project>,
        accountBalanceMap: Map<AccountId, Account>,
    ) {
        val now = Clock.System.now()
        val since = now.minus(term.hours.hours)

        val currentBalance =
            (users.map { it.accountId } + projects.map { it.accountId })
                .distinct()
                .sumOf { accountBalanceMap[it]?.balance ?: 0L }

        // 期間内の取引統計
        val transactionStats = transactionRepository.getStats(since)

        // 期間開始時の残高を推定（現在残高 - 期間内変動）
        val previousBalance = currentBalance - transactionStats.netChange
        val difference = transactionStats.netChange
        val ratio =
            if (previousBalance != 0L) {
                ((difference.toDouble() / previousBalance) * 100).toLong()
            } else {
                0L
            }

        cacheRepository.saveSystemStats(
            term,
            SystemStats(
                balance = currentBalance,
                difference = difference,
                count = transactionStats.count,
                total = transactionStats.total,
                ratio = ratio,
            ),
        )
    }

    private suspend fun updateUsersAggregateStats(
        term: Term,
        users: List<User>,
        accountBalanceMap: Map<AccountId, Account>,
    ) {
        val now = Clock.System.now()
        val since = now.minus(term.hours.hours)

        val currentBalance = users.sumOf { accountBalanceMap[it.accountId]?.balance ?: 0L }
        val transactionStats = transactionRepository.getUsersStats(since)

        val previousBalance = currentBalance - transactionStats.netChange
        val difference = transactionStats.netChange
        val ratio =
            if (previousBalance != 0L) {
                ((difference.toDouble() / previousBalance) * 100).toLong()
            } else {
                0L
            }

        cacheRepository.saveUsersAggregateStats(
            term,
            AggregateStats(
                number = users.size.toLong(),
                balance = currentBalance,
                difference = difference,
                count = transactionStats.count,
                total = transactionStats.total,
                ratio = ratio,
            ),
        )
    }

    private suspend fun updateProjectsAggregateStats(
        term: Term,
        projects: List<Project>,
        accountBalanceMap: Map<AccountId, Account>,
    ) {
        val now = Clock.System.now()
        val since = now.minus(term.hours.hours)

        val currentBalance = projects.sumOf { accountBalanceMap[it.accountId]?.balance ?: 0L }
        val transactionStats = transactionRepository.getProjectsStats(since)

        val previousBalance = currentBalance - transactionStats.netChange
        val difference = transactionStats.netChange
        val ratio =
            if (previousBalance != 0L) {
                ((difference.toDouble() / previousBalance) * 100).toLong()
            } else {
                0L
            }

        cacheRepository.saveProjectsAggregateStats(
            term,
            AggregateStats(
                number = projects.size.toLong(),
                balance = currentBalance,
                difference = difference,
                count = transactionStats.count,
                total = transactionStats.total,
                ratio = ratio,
            ),
        )
    }

    private data class UserStatsData(
        val userId: UserId,
        val balance: Long,
        val difference: Long,
        val inAmount: Long,
        val outAmount: Long,
        val count: Long,
        val total: Long,
        val ratio: Long,
    )

    private data class ProjectStatsData(
        val projectId: ProjectId,
        val balance: Long,
        val difference: Long,
        val inAmount: Long,
        val outAmount: Long,
        val count: Long,
        val total: Long,
        val ratio: Long,
    )

    private suspend fun calculateAllUserRankings(
        term: Term,
        users: List<User>,
        accountBalanceMap: Map<AccountId, Account>,
        previous: Boolean = false,
    ): List<UserStatsData> {
        val now = Clock.System.now()
        val offset = if (previous) term.hours.hours else 0.hours
        val since = now.minus(term.hours.hours + offset)
        val until = if (previous) now.minus(offset) else now

        return users.map { user ->
            val balanceNow = accountBalanceMap[user.accountId]?.balance ?: 0L
            val stats = transactionRepository.getUserStats(user.id, since, until)

            // 期間終了時点の残高を計算
            // previous=true の場合: until時点の残高 = 今の残高 - (untilから今までの変化)
            val balanceAtEnd =
                if (previous) {
                    val changesAfterUntil = transactionRepository.getUserBalanceChangeAfter(user.id, until)
                    balanceNow - changesAfterUntil.netChange
                } else {
                    balanceNow
                }

            // 期間開始時点の残高 = 期間終了時点の残高 - 期間中の変化
            val balanceAtStart = balanceAtEnd - stats.netChange
            val ratio =
                if (balanceAtStart != 0L) {
                    ((stats.netChange.toDouble() / balanceAtStart) * 100).toLong()
                } else {
                    0L
                }

            UserStatsData(
                userId = user.id,
                balance = balanceAtEnd,
                difference = stats.netChange,
                inAmount = stats.inAmount,
                outAmount = stats.outAmount,
                count = stats.count,
                total = stats.total,
                ratio = ratio,
            )
        }
    }

    private suspend fun calculateAllProjectRankings(
        term: Term,
        projects: List<Project>,
        accountBalanceMap: Map<AccountId, Account>,
        previous: Boolean = false,
    ): List<ProjectStatsData> {
        val now = Clock.System.now()
        val offset = if (previous) term.hours.hours else 0.hours
        val since = now.minus(term.hours.hours + offset)
        val until = if (previous) now.minus(offset) else now

        return projects.map { project ->
            val balanceNow = accountBalanceMap[project.accountId]?.balance ?: 0L
            val stats = transactionRepository.getProjectStats(project.id, since, until)

            // 期間終了時点の残高を計算
            val balanceAtEnd =
                if (previous) {
                    val changesAfterUntil = transactionRepository.getProjectBalanceChangeAfter(project.id, until)
                    balanceNow - changesAfterUntil.netChange
                } else {
                    balanceNow
                }

            // 期間開始時点の残高 = 期間終了時点の残高 - 期間中の変化
            val balanceAtStart = balanceAtEnd - stats.netChange
            val ratio =
                if (balanceAtStart != 0L) {
                    ((stats.netChange.toDouble() / balanceAtStart) * 100).toLong()
                } else {
                    0L
                }

            ProjectStatsData(
                projectId = project.id,
                balance = balanceAtEnd,
                difference = stats.netChange,
                inAmount = stats.inAmount,
                outAmount = stats.outAmount,
                count = stats.count,
                total = stats.total,
                ratio = ratio,
            )
        }
    }

    private suspend fun updateUserRankings(
        term: Term,
        rankingType: RankingType,
        currentData: List<UserStatsData>,
        previousData: List<UserStatsData>,
    ) {
        val getValue: (UserStatsData) -> Long = { data ->
            when (rankingType) {
                RankingType.BALANCE -> data.balance
                RankingType.DIFFERENCE -> data.difference
                RankingType.IN -> data.inAmount
                RankingType.OUT -> data.outAmount
                RankingType.COUNT -> data.count
                RankingType.TOTAL -> data.total
                RankingType.RATIO -> data.ratio
            }
        }

        // 現在のランキング（降順）
        val sortedCurrent = currentData.sortedByDescending { getValue(it) }
        val currentRanked = mutableMapOf<UserId, Long>()
        sortedCurrent.forEachIndexed { index, data ->
            if (index > 0 && getValue(data) == getValue(sortedCurrent[index - 1])) {
                currentRanked[data.userId] = currentRanked[sortedCurrent[index - 1].userId]!!
            } else {
                currentRanked[data.userId] = index + 1L
            }
        }

        // 前期間のランキング
        val sortedPrevious = previousData.sortedByDescending { getValue(it) }
        val previousRanked = mutableMapOf<UserId, Long>()
        sortedPrevious.forEachIndexed { index, data ->
            if (index > 0 && getValue(data) == getValue(sortedPrevious[index - 1])) {
                previousRanked[data.userId] = previousRanked[sortedPrevious[index - 1].userId]!!
            } else {
                previousRanked[data.userId] = index + 1L
            }
        }

        val entries =
            currentData.map { data ->
                val currentRank = currentRanked[data.userId] ?: 0L
                val previousRank = previousRanked[data.userId] ?: currentRank
                val rankDifference = previousRank - currentRank // プラスなら順位上昇

                UserRankingEntry(
                    rank = currentRank,
                    value = getValue(data),
                    difference = rankDifference,
                    userId = data.userId,
                )
            }

        cacheRepository.clearUserRankings(term, rankingType)
        cacheRepository.saveUserRankings(term, rankingType, entries)
    }

    private suspend fun updateProjectRankings(
        term: Term,
        rankingType: RankingType,
        currentData: List<ProjectStatsData>,
        previousData: List<ProjectStatsData>,
    ) {
        val getValue: (ProjectStatsData) -> Long = { data ->
            when (rankingType) {
                RankingType.BALANCE -> data.balance
                RankingType.DIFFERENCE -> data.difference
                RankingType.IN -> data.inAmount
                RankingType.OUT -> data.outAmount
                RankingType.COUNT -> data.count
                RankingType.TOTAL -> data.total
                RankingType.RATIO -> data.ratio
            }
        }

        val sortedCurrent = currentData.sortedByDescending { getValue(it) }
        val currentRanked = mutableMapOf<ProjectId, Long>()
        sortedCurrent.forEachIndexed { index, data ->
            if (index > 0 && getValue(data) == getValue(sortedCurrent[index - 1])) {
                currentRanked[data.projectId] = currentRanked[sortedCurrent[index - 1].projectId]!!
            } else {
                currentRanked[data.projectId] = index + 1L
            }
        }

        val sortedPrevious = previousData.sortedByDescending { getValue(it) }
        val previousRanked = mutableMapOf<ProjectId, Long>()
        sortedPrevious.forEachIndexed { index, data ->
            if (index > 0 && getValue(data) == getValue(sortedPrevious[index - 1])) {
                previousRanked[data.projectId] = previousRanked[sortedPrevious[index - 1].projectId]!!
            } else {
                previousRanked[data.projectId] = index + 1L
            }
        }

        val entries =
            currentData.map { data ->
                val currentRank = currentRanked[data.projectId] ?: 0L
                val previousRank = previousRanked[data.projectId] ?: currentRank
                val rankDifference = previousRank - currentRank

                ProjectRankingEntry(
                    rank = currentRank,
                    value = getValue(data),
                    difference = rankDifference,
                    projectId = data.projectId,
                )
            }

        cacheRepository.clearProjectRankings(term, rankingType)
        cacheRepository.saveProjectRankings(term, rankingType, entries)
    }
}

File: internal/service/stats_update_job.go
package service

import (
	"context"
	"log/slog"
	"sort"
	"time"

	"github.com/traP-jp/pteron-server/internal/domain"
)

type StatsCacheWriter interface {
	SaveSystemStats(ctx context.Context, term domain.Term, stats domain.SystemStats) error
	SaveUsersAggregateStats(ctx context.Context, term domain.Term, stats domain.AggregateStats) error
	SaveProjectsAggregateStats(ctx context.Context, term domain.Term, stats domain.AggregateStats) error
	ClearUserRankings(ctx context.Context, term domain.Term, rankingType domain.RankingType) error
	ClearProjectRankings(ctx context.Context, term domain.Term, rankingType domain.RankingType) error
	SaveUserRankings(ctx context.Context, term domain.Term, rankingType domain.RankingType, entries []domain.UserRankingEntry) error
	SaveProjectRankings(ctx context.Context, term domain.Term, rankingType domain.RankingType, entries []domain.ProjectRankingEntry) error
}

type StatsTransactionStore interface {
	GetStats(ctx context.Context, since time.Time) (domain.TransactionStatsData, error)
	GetUsersStats(ctx context.Context, since time.Time) (domain.TransactionStatsData, error)
	GetProjectsStats(ctx context.Context, since time.Time) (domain.TransactionStatsData, error)
	GetUserStats(ctx context.Context, userID domain.UserID, since time.Time, until time.Time) (domain.TransactionStatsData, error)
	GetProjectStats(ctx context.Context, projectID domain.ProjectID, since time.Time, until time.Time) (domain.TransactionStatsData, error)
	GetUserBalanceChangeAfter(ctx context.Context, userID domain.UserID, after time.Time) (domain.BalanceChangeData, error)
	GetProjectBalanceChangeAfter(ctx context.Context, projectID domain.ProjectID, after time.Time) (domain.BalanceChangeData, error)
}

type StatsUpdateJob struct {
	cache interface {
		StatsCacheWriter
	}
	transactions StatsTransactionStore
	users        UserStore
	projects     ProjectStore
	accounts     *AccountService
	logger       *slog.Logger
}

type userStatsData struct {
	UserID     domain.UserID
	Balance    int64
	Difference int64
	InAmount   int64
	OutAmount  int64
	Count      int64
	Total      int64
	Ratio      int64
}

type projectStatsData struct {
	ProjectID  domain.ProjectID
	Balance    int64
	Difference int64
	InAmount   int64
	OutAmount  int64
	Count      int64
	Total      int64
	Ratio      int64
}

func NewStatsUpdateJob(cache StatsCacheWriter, transactions StatsTransactionStore, users UserStore, projects ProjectStore, accounts *AccountService, logger *slog.Logger) *StatsUpdateJob {
	return &StatsUpdateJob{
		cache:        cache,
		transactions: transactions,
		users:        users,
		projects:     projects,
		accounts:     accounts,
		logger:       logger,
	}
}

func (j *StatsUpdateJob) Start(ctx context.Context) {
	go func() {
		timer := time.NewTimer(10 * time.Second)
		defer timer.Stop()
		select {
		case <-ctx.Done():
			return
		case <-timer.C:
		}

		ticker := time.NewTicker(time.Minute)
		defer ticker.Stop()
		for {
			if err := j.UpdateAllStats(ctx); err != nil {
				j.logger.Error("failed to update stats cache", "error", err)
			}
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
			}
		}
	}()
}

func (j *StatsUpdateJob) UpdateAllStats(ctx context.Context) error {
	j.logger.Info("starting stats cache update")
	start := time.Now()

	users, err := j.users.FindAll(ctx)
	if err != nil {
		return err
	}
	projects, err := j.projects.FindAll(ctx)
	if err != nil {
		return err
	}
	accountIDs := make([]domain.AccountID, 0, len(users)+len(projects))
	for _, user := range users {
		accountIDs = append(accountIDs, user.AccountID)
	}
	for _, project := range projects {
		accountIDs = append(accountIDs, project.AccountID)
	}
	accounts, err := j.accounts.GetAccountsByIDs(ctx, uniqueComparable(accountIDs))
	if err != nil {
		return err
	}
	accountByID := make(map[domain.AccountID]domain.Account, len(accounts))
	for _, account := range accounts {
		accountByID[account.ID] = account
	}

	for _, term := range domain.Terms() {
		if err := j.updateTerm(ctx, term, users, projects, accountByID); err != nil {
			j.logger.Error("failed to update stats term", "term", term, "error", err)
		}
	}
	j.logger.Info("stats cache update completed", "elapsed_ms", time.Since(start).Milliseconds())
	return nil
}

func (j *StatsUpdateJob) updateTerm(ctx context.Context, term domain.Term, users []domain.User, projects []domain.Project, accountByID map[domain.AccountID]domain.Account) error {
	now := time.Now().UTC()
	since := now.Add(-time.Duration(term.Hours()) * time.Hour)

	allBalance := int64(0)
	for _, user := range users {
		allBalance += accountByID[user.AccountID].Balance
	}
	for _, project := range projects {
		allBalance += accountByID[project.AccountID].Balance
	}
	systemStats, err := j.transactions.GetStats(ctx, since)
	if err != nil {
		return err
	}
	if err := j.cache.SaveSystemStats(ctx, term, domain.SystemStats{
		Balance: allBalance, Difference: systemStats.NetChange, Count: systemStats.Count, Total: systemStats.Total, Ratio: ratio(systemStats.NetChange, allBalance-systemStats.NetChange),
	}); err != nil {
		return err
	}

	userBalance := int64(0)
	for _, user := range users {
		userBalance += accountByID[user.AccountID].Balance
	}
	userStats, err := j.transactions.GetUsersStats(ctx, since)
	if err != nil {
		return err
	}
	if err := j.cache.SaveUsersAggregateStats(ctx, term, domain.AggregateStats{
		Number: int64(len(users)), Balance: userBalance, Difference: userStats.NetChange, Count: userStats.Count, Total: userStats.Total, Ratio: ratio(userStats.NetChange, userBalance-userStats.NetChange),
	}); err != nil {
		return err
	}

	projectBalance := int64(0)
	for _, project := range projects {
		projectBalance += accountByID[project.AccountID].Balance
	}
	projectStats, err := j.transactions.GetProjectsStats(ctx, since)
	if err != nil {
		return err
	}
	if err := j.cache.SaveProjectsAggregateStats(ctx, term, domain.AggregateStats{
		Number: int64(len(projects)), Balance: projectBalance, Difference: projectStats.NetChange, Count: projectStats.Count, Total: projectStats.Total, Ratio: ratio(projectStats.NetChange, projectBalance-projectStats.NetChange),
	}); err != nil {
		return err
	}

	currentUsers, err := j.calculateAllUserRankings(ctx, term, users, accountByID, false)
	if err != nil {
		return err
	}
	previousUsers, err := j.calculateAllUserRankings(ctx, term, users, accountByID, true)
	if err != nil {
		return err
	}
	currentProjects, err := j.calculateAllProjectRankings(ctx, term, projects, accountByID, false)
	if err != nil {
		return err
	}
	previousProjects, err := j.calculateAllProjectRankings(ctx, term, projects, accountByID, true)
	if err != nil {
		return err
	}

	for _, rankingType := range domain.RankingTypes() {
		if err := j.updateUserRankings(ctx, term, rankingType, currentUsers, previousUsers); err != nil {
			return err
		}
		if err := j.updateProjectRankings(ctx, term, rankingType, currentProjects, previousProjects); err != nil {
			return err
		}
	}
	return nil
}

func (j *StatsUpdateJob) calculateAllUserRankings(ctx context.Context, term domain.Term, users []domain.User, accountByID map[domain.AccountID]domain.Account, previous bool) ([]userStatsData, error) {
	now := time.Now().UTC()
	offset := time.Duration(0)
	if previous {
		offset = time.Duration(term.Hours()) * time.Hour
	}
	since := now.Add(-(time.Duration(term.Hours())*time.Hour + offset))
	until := now.Add(-offset)
	result := make([]userStatsData, 0, len(users))
	for _, user := range users {
		balanceNow := accountByID[user.AccountID].Balance
		stats, err := j.transactions.GetUserStats(ctx, user.ID, since, until)
		if err != nil {
			return nil, err
		}
		balanceAtEnd := balanceNow
		if previous {
			change, err := j.transactions.GetUserBalanceChangeAfter(ctx, user.ID, until)
			if err != nil {
				return nil, err
			}
			balanceAtEnd = balanceNow - change.NetChange()
		}
		result = append(result, userStatsData{
			UserID: user.ID, Balance: balanceAtEnd, Difference: stats.NetChange, InAmount: stats.InAmount, OutAmount: stats.OutAmount, Count: stats.Count, Total: stats.Total, Ratio: ratio(stats.NetChange, balanceAtEnd-stats.NetChange),
		})
	}
	return result, nil
}

func (j *StatsUpdateJob) calculateAllProjectRankings(ctx context.Context, term domain.Term, projects []domain.Project, accountByID map[domain.AccountID]domain.Account, previous bool) ([]projectStatsData, error) {
	now := time.Now().UTC()
	offset := time.Duration(0)
	if previous {
		offset = time.Duration(term.Hours()) * time.Hour
	}
	since := now.Add(-(time.Duration(term.Hours())*time.Hour + offset))
	until := now.Add(-offset)
	result := make([]projectStatsData, 0, len(projects))
	for _, project := range projects {
		balanceNow := accountByID[project.AccountID].Balance
		stats, err := j.transactions.GetProjectStats(ctx, project.ID, since, until)
		if err != nil {
			return nil, err
		}
		balanceAtEnd := balanceNow
		if previous {
			change, err := j.transactions.GetProjectBalanceChangeAfter(ctx, project.ID, until)
			if err != nil {
				return nil, err
			}
			balanceAtEnd = balanceNow - change.NetChange()
		}
		result = append(result, projectStatsData{
			ProjectID: project.ID, Balance: balanceAtEnd, Difference: stats.NetChange, InAmount: stats.InAmount, OutAmount: stats.OutAmount, Count: stats.Count, Total: stats.Total, Ratio: ratio(stats.NetChange, balanceAtEnd-stats.NetChange),
		})
	}
	return result, nil
}

func (j *StatsUpdateJob) updateUserRankings(ctx context.Context, term domain.Term, rankingType domain.RankingType, current []userStatsData, previous []userStatsData) error {
	currentRanks := ranks(current, func(data userStatsData) domain.ID { return data.UserID }, func(data userStatsData) int64 { return userRankingValue(data, rankingType) })
	previousRanks := ranks(previous, func(data userStatsData) domain.ID { return data.UserID }, func(data userStatsData) int64 { return userRankingValue(data, rankingType) })
	entries := make([]domain.UserRankingEntry, 0, len(current))
	for _, data := range current {
		currentRank := currentRanks[data.UserID]
		previousRank := previousRanks[data.UserID]
		if previousRank == 0 {
			previousRank = currentRank
		}
		entries = append(entries, domain.UserRankingEntry{Rank: currentRank, Value: userRankingValue(data, rankingType), Difference: previousRank - currentRank, UserID: data.UserID})
	}
	if err := j.cache.ClearUserRankings(ctx, term, rankingType); err != nil {
		return err
	}
	return j.cache.SaveUserRankings(ctx, term, rankingType, entries)
}

func (j *StatsUpdateJob) updateProjectRankings(ctx context.Context, term domain.Term, rankingType domain.RankingType, current []projectStatsData, previous []projectStatsData) error {
	currentRanks := ranks(current, func(data projectStatsData) domain.ID { return data.ProjectID }, func(data projectStatsData) int64 { return projectRankingValue(data, rankingType) })
	previousRanks := ranks(previous, func(data projectStatsData) domain.ID { return data.ProjectID }, func(data projectStatsData) int64 { return projectRankingValue(data, rankingType) })
	entries := make([]domain.ProjectRankingEntry, 0, len(current))
	for _, data := range current {
		currentRank := currentRanks[data.ProjectID]
		previousRank := previousRanks[data.ProjectID]
		if previousRank == 0 {
			previousRank = currentRank
		}
		entries = append(entries, domain.ProjectRankingEntry{Rank: currentRank, Value: projectRankingValue(data, rankingType), Difference: previousRank - currentRank, ProjectID: data.ProjectID})
	}
	if err := j.cache.ClearProjectRankings(ctx, term, rankingType); err != nil {
		return err
	}
	return j.cache.SaveProjectRankings(ctx, term, rankingType, entries)
}

func ranks[T any](items []T, id func(T) domain.ID, value func(T) int64) map[domain.ID]int64 {
	sorted := append([]T(nil), items...)
	sort.SliceStable(sorted, func(i, j int) bool { return value(sorted[i]) > value(sorted[j]) })
	result := make(map[domain.ID]int64, len(sorted))
	for i, item := range sorted {
		if i > 0 && value(item) == value(sorted[i-1]) {
			result[id(item)] = result[id(sorted[i-1])]
		} else {
			result[id(item)] = int64(i + 1)
		}
	}
	return result
}

func userRankingValue(data userStatsData, rankingType domain.RankingType) int64 {
	switch rankingType {
	case domain.RankingBalance:
		return data.Balance
	case domain.RankingDifference:
		return data.Difference
	case domain.RankingIn:
		return data.InAmount
	case domain.RankingOut:
		return data.OutAmount
	case domain.RankingCount:
		return data.Count
	case domain.RankingTotal:
		return data.Total
	case domain.RankingRatio:
		return data.Ratio
	default:
		return 0
	}
}

func projectRankingValue(data projectStatsData, rankingType domain.RankingType) int64 {
	switch rankingType {
	case domain.RankingBalance:
		return data.Balance
	case domain.RankingDifference:
		return data.Difference
	case domain.RankingIn:
		return data.InAmount
	case domain.RankingOut:
		return data.OutAmount
	case domain.RankingCount:
		return data.Count
	case domain.RankingTotal:
		return data.Total
	case domain.RankingRatio:
		return data.Ratio
	default:
		return 0
	}
}

func ratio(difference int64, previousBalance int64) int64 {
	if previousBalance == 0 {
		return 0
	}
	return int64((float64(difference) / float64(previousBalance)) * 100)
}

func uniqueComparable[T comparable](values []T) []T {
	seen := make(map[T]struct{}, len(values))
	out := make([]T, 0, len(values))
	for _, value := range values {
		if _, ok := seen[value]; ok {
			continue
		}
		seen[value] = struct{}{}
		out = append(out, value)
	}
	return out
}
